providers/base/tests/test_network_device_info.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+# Copyright 2020 Canonical Ltd.
+# Written by:
+#   Dio He <dio.he@canonical.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3,
+# as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import unittest
+from unittest.mock import Mock, patch
+
+from network_device_info import Utils
+
+class GetIpv4AddressTests(unittest.TestCase):
+    # There are 2 output we could get from fcntl.ioctl():
+    # 1. No WiFi connected 
+    #    (It pulls out bytes data directly from kernel, so if the given SIOCGIFADDR doesn't exist, it raise a OSError)
+
+    # 2. Connected with WiFi
+    #    It will pulls out a 256 bytes data into buffer, then we get to parse the IP address
+    @patch("fcntl.ioctl")
+    def test_get_ipv4_address_without_connection(self, mock_ioctl):
+        mock_ioctl.side_effect = OSError()
+        interface = "wlo1"
+        addr = Utils.get_ipv4_address(interface)
+        self.assertEqual(addr, "***NOT CONFIGURED***")
```

Comment:
In order to avoid the error printing as an output of the tests, you could use [Trey Hunter's suggestion](https://www.reddit.com/r/learnpython/comments/5f0vko/comment/dago5w7/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) of using `redirect_stdout` (except you would probably want to catch `stderr` instead using [`redirect_stderr`](https://docs.python.org/3/library/contextlib.html#contextlib.redirect_stderr)).

Reply:
Thank you very much for the link!
what do you mean by catch stderr? I thought calling redirect_stderr() to stop from print it out is the solution?

Reply:
Just to clarify, we only want to know if the function could pass the unit test or not, right?
Or error msg should be printed if the test is failed? (other existing unit tests don't print anything)